src/mapping/PartitionOfUnityMapping.hpp

```diff
@@ -0,0 +1,390 @@
+  /// only enabled in debug builds and mainly for debugging purpose
+  void exportClusterCentersAsVTU(mesh::Mesh &centers);
+};
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+PartitionOfUnityMapping<RADIAL_BASIS_FUNCTION_T>::PartitionOfUnityMapping(
+    Mapping::Constraint constraint,
+    int                 dimension,
+    double              parameter,
+    Polynomial          polynomial,
+    unsigned int        verticesPerCluster,
+    double              relativeOverlap,
+    bool                projectToInput)
+    : Mapping(constraint, dimension),
+      _parameter(parameter), _verticesPerCluster(verticesPerCluster), _relativeOverlap(relativeOverlap), _projectToInput(projectToInput), _polynomial(polynomial)
+{
+  PRECICE_ASSERT(this->getDimensions() <= 3);
+  PRECICE_ASSERT(_relativeOverlap < 1, "The relative overlap has to be smaller than one.");
+  PRECICE_ASSERT(_verticesPerCluster > 0, "The number of vertices per cluster has to be greater zero.");
+
+  if (isScaledConsistent()) {
+    setInputRequirement(Mapping::MeshRequirement::FULL);
+    setOutputRequirement(Mapping::MeshRequirement::FULL);
+  } else {
+    setInputRequirement(Mapping::MeshRequirement::VERTEX);
+    setOutputRequirement(Mapping::MeshRequirement::VERTEX);
+  }
+  // Simply set each axis to be active according to the space dimension.
+  // We could even set each axis to be true, which would just lead to 'zero' columns in the polynomial QR.
+  // Integrated polynomials are anyway not supported with PoU
+  _deadAxis.clear();
+  std::fill_n(std::back_inserter(_deadAxis), getDimensions(), false);
+}
+
+template <typename RADIAL_BASIS_FUNCTION_T>
+void PartitionOfUnityMapping<RADIAL_BASIS_FUNCTION_T>::computeMapping()
+{
+  PRECICE_TRACE();
+
+  precice::utils::Event e("map.pou.computeMapping.From" + this->input()->getName() + "To" + this->output()->getName(), precice::syncMode);
+
+  // Recompute the whole clustering
+  PRECICE_ASSERT(!this->_hasComputedMapping, "Please clear the mapping before recomputing.");
+
+  mesh::PtrMesh inMesh;
+  mesh::PtrMesh outMesh;
+  if (this->hasConstraint(Mapping::CONSERVATIVE)) {
+    inMesh  = this->output();
+    outMesh = this->input();
+  } else { // Consistent or scaled consistent
+    inMesh  = this->input();
+    outMesh = this->output();
+  }
+
+  // Step 1: get a tentative clustering consisting of centers and a radius from one of the available algorithms
+  auto [averageClusterRadius_, centerCandidates] = impl::createClustering(inMesh, outMesh, _relativeOverlap, _verticesPerCluster, _projectToInput);
+
+  averageClusterRadius = averageClusterRadius_;
+  PRECICE_ASSERT(averageClusterRadius > 0 || inMesh->vertices().size() == 0 || outMesh->vertices().size() == 0);
+
+  // Step 2: check, which of the resulting clusters are non-empty and register the cluster centers in a mesh
+  // Here, the VertexCluster computes the matrix decompositions directly in case the cluster is non-empty
```

Comment:
We already check for empty-ness in the `createClustering`. We might, however, 'project' the centers afterwards and remove duplicates from the center vertices. In case we don't do the projection, we could in principle merge things from the cluster creation and the index queries here, but it makes the code more complicated. I would keep it this way.